### PR TITLE
Improve typings for event emitters

### PR DESCRIPTION
**Pull request by unao**
As a logical continuation of https://github.com/versatica/mediasoup/pull/764 would like to add analogous types for mediasoup client.

I introduced some _implicit_ assumption around `safeEmitAsPromise`: if corresponding type is not specified it allows to pass any arguments otherwise it must be a single argument conforming to the first declared type.

There is still some room left for improvements around `AppData` (for preventing unsafe cast) - if you were interested I could add it once this is merged in.

### CODE REVIEW

**Review by jmillan — APPROVED** [review]
:+1:

**Comment by unao** [comment]
@jmillan @ibc I resolved some conflicts in generated types. Any chance to merge this in?

**Comment by ibc** [comment]
Will review soon, sorry for the delay.

**Comment by unao** [comment]
@ibc sure no hurry ;) 
btw. I resolved some other conflicts that popped up since then

**Comment by ibc** [comment]
We don't forget about this. It's just that we are too busy.

**Comment by unao** [comment]
I probably will close this PR soon and just fork this repo to move forward.

**Comment by ibc** [comment]
Please don't. Will manage this PR this week. Promise.

File: src/Transport.ts [hunk 1]
```diff
@@ -148,9 +149,41 @@ export type PlainRtpParameters =
 	port: number;
 };
 
+export type TransportProduceParameters<AppData = unknown> =
```
Comment (ibc) [inline src/Transport.ts:152]: `AppData` default value should not be `unkown` but `Record<string, unknown>` (everywhere).

**Review by ibc — CHANGES_REQUESTED** [review]

**Comment by ibc** [comment]
@unao I've made some changes that improve `EnhancedEventEmitter` and I assume we should use the same in mediasoup-client:

https://github.com/versatica/mediasoup/pull/836

**Comment by ibc** [comment]
Closed in favour of https://github.com/versatica/mediasoup-client/pull/209

**Pull request closed without merging**